File: frontend/src/utils/idleRate.js
const PERIOD_COUNT = 14;

const CLASS_START_TIME = [
  "08:00",
  "08:50",
  "09:50",
  "10:40",
  "11:30",
  "13:00",
  "13:50",
  "14:45",
  "15:40",
  "16:35",
  "17:25",
  "18:30",
  "19:20",
  "20:10",
];

const CLASS_END_TIME = [
  "08:45",
  "09:35",
  "10:35",
  "11:25",
  "12:15",
  "13:45",
  "14:35",
  "15:30",
  "16:25",
  "17:20",
  "18:10",
  "19:15",
  "20:05",
  "20:55",
];

export function getSelectedBuildingNames(campusInfo, selectedBuildings) {
  if (!campusInfo || !Array.isArray(selectedBuildings) || selectedBuildings.length === 0) {
    return [];
  }

  const idToName = new Map();
  for (const [buildingName, buildingId] of Object.entries(campusInfo.building_id_map ?? {})) {
    idToName.set(String(buildingId), buildingName);
  }

  return selectedBuildings
    .map((buildingId) => idToName.get(String(buildingId)))
    .filter(Boolean);
}

export function calculateTodayIdleRate({
  classInfo,
  selectedCampus,
  selectedBuildings,
}) {
  const campusInfo = classInfo?.campus_info_map?.[selectedCampus];
  const catalog = classInfo?.classroom_catalog?.[selectedCampus];
  const buildingNames = getSelectedBuildingNames(campusInfo, selectedBuildings);
  if (!campusInfo || !catalog || buildingNames.length === 0) {
    return [];
  }

  const totals = Array.from({ length: PERIOD_COUNT }, (_, periodIndex) => ({
    periodIndex,
    label: String(periodIndex + 1).padStart(2, "0"),
    startTime: CLASS_START_TIME[periodIndex],
    endTime: CLASS_END_TIME[periodIndex],
    timeRange: `${CLASS_START_TIME[periodIndex]}-${CLASS_END_TIME[periodIndex]}`,
    emptyClassrooms: 0,
    totalClassrooms: 0,
    idleRate: 0,
  }));

  for (const buildingName of buildingNames) {
    const rooms = catalog[buildingName];
    if (!rooms) {
      continue;
    }

    const roomNames = Object.entries(rooms)
      .map(([roomName, entry]) => entry?.classroom_name || roomName)
      .filter(Boolean);
    if (roomNames.length === 0) {
      continue;
    }

    const buildingInfo = findBuildingInfo(campusInfo, buildingName);
    for (let periodIndex = 0; periodIndex < PERIOD_COUNT; periodIndex += 1) {
      let emptyClassrooms = 0;
      for (const roomName of roomNames) {
        if (isClassroomEmptyAtPeriod(buildingInfo, roomName, periodIndex)) {
          emptyClassrooms += 1;
        }
      }
      totals[periodIndex].emptyClassrooms += emptyClassrooms;
      totals[periodIndex].totalClassrooms += roomNames.length;
    }
  }

  const hasKnownClassrooms = totals.some((period) => period.totalClassrooms > 0);
  if (!hasKnownClassrooms) {
    return [];
  }

  return totals.map((period) => ({
    ...period,
    idleRate:
      period.totalClassrooms === 0
        ? 0
        : period.emptyClassrooms / period.totalClassrooms,
  }));
}

export function calculateWeeklyIdleRate({
  idleRateHistory,
  selectedCampus,
  selectedBuildings,
  campusInfo,
}) {
  const buildingNames = getSelectedBuildingNames(campusInfo, selectedBuildings);
  if (!Array.isArray(idleRateHistory) || buildingNames.length === 0) {
    return [];
  }

  return idleRateHistory
    .slice()
    .sort((a, b) => String(a.date).localeCompare(String(b.date)))
    .map((entry) => {
      const campusStats = entry?.campus?.[selectedCampus];
      if (!campusStats) {
        return null;
      }

      let emptyClassroomPeriods = 0;
      let totalClassroomPeriods = 0;
      for (const buildingName of buildingNames) {
        const buildingStats = campusStats[buildingName];
        if (!buildingStats) {
          continue;
        }

        for (const period of buildingStats.periods ?? []) {
          emptyClassroomPeriods += period.empty_classrooms ?? 0;
          totalClassroomPeriods +=
            period.total_classrooms ?? buildingStats.known_classrooms ?? 0;
        }
      }

      if (totalClassroomPeriods === 0) {
        return null;
      }

      return {
        date: entry.date,
        averageRate: emptyClassroomPeriods / totalClassroomPeriods,
        averageEmptyClassrooms: emptyClassroomPeriods / PERIOD_COUNT,
        totalClassrooms: totalClassroomPeriods / PERIOD_COUNT,
      };
    })
    .filter(Boolean);
}

function findBuildingInfo(campusInfo, buildingName) {
  const buildingId = campusInfo?.building_id_map?.[buildingName];
  if (buildingId === undefined || buildingId === null) {
    return null;
  }
  return campusInfo?.building_info_map?.[buildingId] ?? null;
}

function isClassroomEmptyAtPeriod(buildingInfo, roomName, periodIndex) {
  if (!buildingInfo) {
    return false;
  }
  const classroomId = buildingInfo.classroom_id_map?.[roomName];
  if (classroomId === undefined || classroomId === null) {
    return false;
  }
  return buildingInfo.class_matrix?.[periodIndex]?.[classroomId] === 0;
}

File: service/idle_rate.go
package service

import (
	"EmptyClassroom/service/model"
	"sort"
	"time"
)

const idleRateHistoryLimit = 7
const classPeriodsPerDay = 14

func MergeIdleRateHistory(previous *model.ClassInfo, current *model.ClassInfo, observedAt time.Time) {
	if current == nil {
		return
	}

	var previousHistory []model.IdleRateHistoryEntry
	if previous != nil {
		previousHistory = previous.IdleRateHistory
	}
	currentEntry := buildIdleRateHistoryEntry(current, observedAt)
	current.IdleRateHistory = upsertIdleRateHistory(previousHistory, currentEntry)
}

func buildIdleRateHistoryEntry(classInfo *model.ClassInfo, observedAt time.Time) model.IdleRateHistoryEntry {
	entry := model.IdleRateHistoryEntry{
		Date:   formatIdleRateDate(observedAt),
		Campus: map[string]model.IdleRateCampusStats{},
	}
	if classInfo == nil {
		return entry
	}

	for campusName, buildings := range classInfo.ClassroomCatalog {
		campusStats := model.IdleRateCampusStats{}
		for buildingName, rooms := range buildings {
			buildingStats := buildIdleRateBuildingStats(classInfo, campusName, buildingName, rooms)
			if buildingStats.KnownClassrooms == 0 {
				continue
			}
			campusStats[buildingName] = buildingStats
		}
		if len(campusStats) > 0 {
			entry.Campus[campusName] = campusStats
		}
	}

	return entry
}

func buildIdleRateBuildingStats(classInfo *model.ClassInfo, campusName string, buildingName string, rooms model.RoomClassroomCatalog) model.IdleRateBuildingStats {
	stats := model.IdleRateBuildingStats{
		Periods: make([]model.IdleRatePeriodStat, classPeriodsPerDay),
	}

	roomNames := make([]string, 0, len(rooms))
	for roomName, room := range rooms {
		if room == nil {
			continue
		}
		catalogRoomName := room.ClassroomName
		if catalogRoomName == "" {
			catalogRoomName = roomName
		}
		if catalogRoomName == "" {
			continue
		}
		roomNames = append(roomNames, catalogRoomName)
	}
	stats.KnownClassrooms = len(roomNames)
	if stats.KnownClassrooms == 0 {
		return stats
	}

	buildingInfo := findBuildingInfo(classInfo, campusName, buildingName)
	rateSum := 0.0
	for periodIndex := 0; periodIndex < classPeriodsPerDay; periodIndex++ {
		emptyClassrooms := 0
		for _, roomName := range roomNames {
			if isClassroomEmptyAtPeriod(buildingInfo, roomName, periodIndex) {
				emptyClassrooms++
			}
		}
		idleRate := float64(emptyClassrooms) / float64(stats.KnownClassrooms)
		stats.Periods[periodIndex] = model.IdleRatePeriodStat{
			EmptyClassrooms: emptyClassrooms,
			TotalClassrooms: stats.KnownClassrooms,
			IdleRate:        idleRate,
		}
		rateSum += idleRate
	}
	stats.AverageRate = rateSum / classPeriodsPerDay
	return stats
}

func findBuildingInfo(classInfo *model.ClassInfo, campusName string, buildingName string) *model.BuildingInfo {
	if classInfo == nil || classInfo.CampusInfoMap == nil {
		return nil
	}
	campusInfo := classInfo.CampusInfoMap[campusName]
	if campusInfo == nil {
		return nil
	}
	buildingID, ok := campusInfo.BuildingIdMap[buildingName]
	if !ok {
		return nil
	}
	return campusInfo.BuildingInfoMap[buildingID]
}

func isClassroomEmptyAtPeriod(buildingInfo *model.BuildingInfo, roomName string, periodIndex int) bool {
	if buildingInfo == nil || periodIndex < 0 || periodIndex >= len(buildingInfo.ClassMatrix) {
		return false
	}
	classroomID, ok := buildingInfo.ClassroomIdMap[roomName]
	if !ok {
		return false
	}
	period := buildingInfo.ClassMatrix[periodIndex]
	if classroomID < 0 || classroomID >= len(period) {
		return false
	}
	return period[classroomID] == 0
}

func upsertIdleRateHistory(previous []model.IdleRateHistoryEntry, current model.IdleRateHistoryEntry) []model.IdleRateHistoryEntry {
	byDate := map[string]model.IdleRateHistoryEntry{}
	for _, entry := range previous {
		if entry.Date == "" {
			continue
		}
		byDate[entry.Date] = entry
	}
	if current.Date != "" {
		byDate[current.Date] = current
	}

	dates := make([]string, 0, len(byDate))
	for date := range byDate {
		dates = append(dates, date)
	}
	sort.Strings(dates)
	if len(dates) > idleRateHistoryLimit {
		dates = dates[len(dates)-idleRateHistoryLimit:]
	}

	history := make([]model.IdleRateHistoryEntry, 0, len(dates))
	for _, date := range dates {
		history = append(history, byDate[date])
	}
	return history
}

func formatIdleRateDate(observedAt time.Time) string {
	location, err := time.LoadLocation("Asia/Shanghai")
	if err != nil {
		location = time.FixedZone("Asia/Shanghai", 8*60*60)
	}
	return observedAt.In(location).Format("2006-01-02")
}
